src/main/java/com/robertx22/mine_and_slash/a_libraries/dmg_number_particle/particle/style/Row.java

```diff
@@ -0,0 +1,122 @@
+package com.robertx22.mine_and_slash.a_libraries.dmg_number_particle.particle.style;
+
+import com.mojang.blaze3d.vertex.PoseStack;
+import com.mojang.blaze3d.vertex.VertexConsumer;
+import com.mojang.math.Axis;
+import com.robertx22.mine_and_slash.a_libraries.dmg_number_particle.particle.ExileInteractionResultParticle;
+import com.robertx22.mine_and_slash.a_libraries.dmg_number_particle.particle.GLUtils;
+import net.minecraft.ChatFormatting;
+import net.minecraft.client.Camera;
+import net.minecraft.client.Minecraft;
+import net.minecraft.client.gui.Font;
+import net.minecraft.client.renderer.MultiBufferSource;
+import net.minecraft.network.chat.MutableComponent;
+import net.minecraft.network.chat.contents.LiteralContents;
+import net.minecraft.util.Mth;
+import net.minecraft.world.phys.Vec3;
+import org.apache.commons.lang3.tuple.Pair;
+
+import java.util.ArrayList;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class Row implements IParticleRenderStrategy {
+
+    @Override
+    public void setupStyle(ExileInteractionResultParticle particle) {
+        particle.setGravity(0.0f);
+        particle.setParticleSpeed(0, 0.06f, 0);
+        particle.setLiftTime(20);
+    }
+
+    @Override
+    public void tick(ExileInteractionResultParticle particle) {
+
+    }
+
+    @Override
+    public void setupParticle(ExileInteractionResultParticle particle, VertexConsumer vertexConsumer, Camera camera, float partialTick, PoseStack posestack) {
+        Vec3 cameraPos = camera.getPosition();
+        Vec3 pos = particle.getPos();
+        Vec3 original = particle.getOriginalPosition();
+        double x = (float) (Mth.lerp((double) partialTick, original.x, pos.x));
+        double y = (float) (Mth.lerp((double) partialTick, original.y, pos.y));
+        double z = (float) (Mth.lerp((double) partialTick, original.z, pos.z));
+        float scale = 0.02f;
+        posestack.translate(x - cameraPos.x, y - cameraPos.y, z - cameraPos.z);
+        posestack.mulPose(camera.rotation());
+        posestack.mulPose(Axis.ZP.rotationDegrees(180.0F));
+        posestack.scale(scale, scale, scale);
+        posestack.translate(0.0F, -2.0F, 0.0F);
+    }
+
+    @Override
+    public void renderDamage(ExileInteractionResultParticle particle, VertexConsumer vertexConsumer, Camera camera, float partialTick, PoseStack posestack, String text, int color) {
+        MultiBufferSource.BufferSource multibuffersource$buffersource = Minecraft.getInstance().renderBuffers().bufferSource();
+        boolean isCrit = text.contains("!");
```

Comment:
I'll probably merge it but this here is the sole reason i'd completely rework the entire thing.
Why not just pass the particle data so you don't have to do regex magic to figure out how to render it?